### DISCUSSION TITLE
Minor developer experience kvetch with function names

### DISCUSSION THREAD

**Discussion by JJJ** [spatie/ray#164] [Ideas]
I've been using Ray for a week, and so far have found it to be an enjoyable addition to my workflow. So first, thanks to everyone who is helping improve it!

The nice thing about `var_dump()` as a global function, is that I can easily search for it in my codebase before committing and pushing, or even setup pre-commit hooks for it to warn me if I'm about to commit some debugging stuff that doesn't belong in the commit history.

The problem, then, with `ray()` as a global function (and the new `rd()`) is that they fuzzy match unintended and very commonly named functions and methods like `array(`, `_password(` or `_dashboard(`.

In my projects, I've included these global functions as aliases:

```php
function __r( ...$args ) {
	ray($args);
}

function __rd( ...$args ) {
	rd($args);
}
```

This way, I can stay on the look-out for `__r(` and `__rd(` and make sure my debugging approach does not accidentally get pushed into an environment it doesn't belong in.

(The only problem with this approach is that the link to the local file is now my alias instead of the actual caller.)

**Comment by freekmurze** [comment]
Hi,

thanks for your kind words on Ray and this feedback.

You mention valid drawback on the name. We're not going to change them anytime soon (that would be a breaking change), and not add new ones (because we don't want to pollute the global scope with aliases functions).

Personally I've not missed the ability to search these statements too much, as a link to them is displayed in Ray.

**Reply by JJJ** [reply]
Hello!

All understandable perspectives. If Ray were a project of mine I believe I would feel similarly.

It is a minor inconvenience for me as well, so there is no urgency to consider changing or fixing anything. Mostly I wanted to share that experience early so that others could ponder it with me.

The only change I can envision being useful to others in my situation comes in the form of a `ray.php` configuration option, that would allow for offsetting the debug backtrace (unshift) of the calling function by `1` or more, with `0` as the default. Such an option could, hypothetically, allow anyone to alias Ray to their needs inside their environments or applications within any number of functions or methods, and still get the correct caller sent back with it.

I'm not so much asking for this change, as I am politely sharing my own musings on the problem I'm creating for myself.

Thank you again for your kind response 🙏

**Reply by vesper8** [reply]
@JJJ @freekmurze I have actually reached the exact same situation. I created my own global function which I called `roy`. Both because it's easier to search for, but also because I wanted the means of setting a global colour. So I ended up with this

```
<?php

if (!function_exists('roy')) {

    function roy(...$v)
    {
        ray($v)->purple();
    }
}

if (!function_exists('royd')) {

    function royd(...$v)
    {
        ray($v)->purple();
        exit;
    }
}

```

The problem here is that in Ray, the "caller" is always the file where I've set this global function and not the actual caller.

After digging a bit deeper into how ray determines the backtrace, I found that if I add this line

```
$indexOfRay += 2;
```

At line 52 here https://github.com/spatie/laravel-ray/blob/main/src/OriginFactory.php#L52

Then it shows the correct caller despite my global hack.

@freekmurze Any chance you might allow us to configure this backtrace offset through the configuration file?

Thanks!

**Reply by vesper8** [reply]
Forked laravel-ray https://github.com/vesper8/laravel-ray/blob/main/src/OriginFactory.php

Made backtrace offset configurable

```
$indexOfRay += config('ray.backtrace_offset', 0);
```

config/ray.php
```
<?php

return [
    'backtrace_offset' => 2,
];
```


Still hoping you add this in so I don't have to keep using a fork.

Cheers!

**Comment by dillingham** [comment]
@JJJ here is some regex you can search with: `\s(ray\(?.+\))`

**Reply by JJJ** [reply]
Hi @dillingham and thank you for this suggestion. I appreciate you sharing it here so that myself and others who may find this discussion have a search string that should work inside most editors and environments. 🙏
